File: 2023/08.24/송관재/벼락치기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int t = Integer.parseInt(st.nextToken());
        int chap[][] = new int[n+1][2];//0 time, 1 score
        for(int i=1; i<=n; i++) {
            st = new StringTokenizer(br.readLine());
            for(int j=0; j<2; j++) {
                chap[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        int dp[][] = new int[n+1][t+1];
        for(int i=1; i<=n; i++) {
            for(int j=0; j<=t; j++) {
                if(chap[i][0] <= j) {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - chap[i][0]] + chap[i][1]);
                } else {
                    dp[i][j] = dp[i-1][j];
                }
            }
        }
        System.out.println(dp[n][t]);
    }

}

File: 2023/08.24/조수연/벼락치기.kt
import java.io.*
import java.util.*
import kotlin.math.max

lateinit var st: StringTokenizer

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val (n, t) = readLine().split(" ").map { it.toInt() }
    val scores = Array(n + 1) { IntArray(2) }

    repeat(n) {
        st = StringTokenizer(readLine())
        scores[it + 1][0] = st.nextToken().toInt()
        scores[it + 1][1] = st.nextToken().toInt()
    }

    val dp = Array(n + 1) { IntArray(t + 1) }

    // knapsack
    for (i in 1..n) {
        for (j in 1..t) {
            if (scores[i][0] > j) {
                dp[i][j] = dp[i - 1][j]
            } else {
                dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - scores[i][0]] + scores[i][1])
            }
        }
    }

    bw.write("${dp[n][t]}")

    bw.flush()
    bw.close()
}